src/adapter/src/rbac.rs

```diff
@@ -611,47 +633,93 @@ fn generate_required_privileges(
             let mut privileges =
                 generate_read_privileges(catalog, depends_on.iter().cloned(), role_id);
             if let Ok(cluster) = catalog.resolve_cluster(None) {
-                privileges.insert((cluster.id().into(), AclMode::USAGE, role_id));
+                privileges.push((cluster.id().into(), AclMode::USAGE, role_id));
             }
             privileges
         }
         Plan::Explain(_) => generate_read_privileges(catalog, depends_on.iter().cloned(), role_id),
-        Plan::Insert(plan) => {
+        Plan::CopyFrom(plan) => {
             let item = catalog.get_item(&plan.id);
-            BTreeSet::from([
+            vec![
                 (
                     item.name().qualifiers.clone().into(),
                     AclMode::USAGE,
                     role_id,
                 ),
                 (plan.id.into(), AclMode::INSERT, role_id),
-            ])
+            ]
         }
-        Plan::CopyFrom(plan) => {
-            let item = catalog.get_item(&plan.id);
-            BTreeSet::from([
-                (
-                    item.name().qualifiers.clone().into(),
-                    AclMode::USAGE,
-                    role_id,
-                ),
+        Plan::Insert(plan) => {
+            let update_schema_id: ObjectId =
+                catalog.get_item(&plan.id).name().qualifiers.clone().into();
+            let mut privileges = vec![
+                (update_schema_id.clone(), AclMode::USAGE, role_id),
                 (plan.id.into(), AclMode::INSERT, role_id),
-            ])
+            ];
+            // plan.values has not been optimized yet, so it might be constant even if
+            // `plan.values.as_const().is_none()` is true.
+            privileges.extend_from_slice(&generate_read_privileges_inner(
+                catalog,
+                plan.values.depends_on().into_iter(),
+                role_id,
+                &mut BTreeSet::from([(update_schema_id, role_id)]),
+            ));
+            // Collect any USAGE privileges from the returning clause.
+            let seen = privileges
+                .iter()
+                .filter_map(|(id, _, _)| match id {
+                    ObjectId::Item(id) => Some(*id),
+                    _ => None,
+                })
+                .collect();
+            privileges.extend(generate_item_usage_privileges_inner(
+                catalog, depends_on, role_id, seen,
+            ));
+            // values hasn't been fully optimized yet, so it might actually be a constant,
```

Comment:
This class of problem should also maybe be a result of some function? `plan_uses_cluster(&self: Plan) -> bool` kinda deal? Not sure if this is a good idea or not.

Reply:
I created `generate_cluster_usage_privileges` let me know if that works.